src/Microsoft.Kiota.Cli.Commons/Http/Headers/InMemoryHeadersStore.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Kiota.Cli.Commons.Http.Headers;
+
+/// <summary>
+/// An in memory headers store.
+/// </summary>
+/// 
+/// <remarks>
+/// <para>
+/// Use <see cref="Instance"/> static property to get a default thread unsafe
+/// header store and <see cref="ConcurrentInstance"/> static property to get a
+/// thread-safe variant.
+/// </para>
+/// <para>
+/// Use the <see cref="IsConcurrent"/> instance property to determine if an
+/// instance of this class is thread-safe.
+/// </para>
+/// </remarks>
+public sealed class InMemoryHeadersStore : BaseHeadersStore
+{
+    private readonly IDictionary<string, ICollection<string>> _headers;
+
+    /// <summary>
+    /// Returns a thread unsafe single instance of the
+    /// <see cref="InMemoryHeadersStore"/> singleton.
+    /// </summary>
+    /// 
+    /// <remarks>
+    /// For cases where you might not want to use a single instance across your
+    /// application e.g. when using an instance per thread, create a new
+    /// instance and use that instead. In that case, make sure you have cached
+    /// the instance when you need to access the stored headers.
+    /// </remarks>
+    public static InMemoryHeadersStore Instance { get; } = new();
+
+    /// <summary>
+    /// Returns a thread safe single instance of the
+    /// <see cref="InMemoryHeadersStore"/> singleton.
+    /// </summary>
+    public static InMemoryHeadersStore ConcurrentInstance { get; } = new(true);
+
+    /// <summary>
+    /// A boolean indicating whether the store can be used in multi-threaded
+    /// contexts safely.
+    /// </summary>
+    public bool IsConcurrent { get; }
+
+    /// <inheritdoc />
+    public override IEnumerable<KeyValuePair<string, ICollection<string>>> GetHeaders() => _headers.AsEnumerable();
+
+    /// <inheritdoc />
+    public override IEnumerable<KeyValuePair<string, ICollection<string>>> Drain()
+    {
+        var existing = _headers.ToList();
+        _headers.Clear();
+        return existing;
+    }
+
+    /// <inheritdoc />
+    protected override void DoAddHeader(string name, string value)
+    {
+        if (!_headers.TryAdd(name, new List<string> { value }))
+        {
+            _headers[name].Add(value);
+        }
+    }
+
+    /// <summary>
+    /// Creates a new instance of <see cref="InMemoryHeadersStore"/>
+    /// </summary>
+    /// 
+    /// <param name="concurrent">Whether to opt into thread safety.</param>
+    /// 
+    /// <remarks>
+    /// When using the headers store from multiple threads, you can opt into
+    /// thread synchronization by setting <paramref name="concurrent"/> to
+    /// true. This option uses a thread safe backing store, but will incur
+    /// some synchronization overhead.
+    /// </remarks>
+    public InMemoryHeadersStore(bool concurrent = false)
+    {
+        if (concurrent)

```

Comment:
There's a performance impact to using a concurrent dictionary that I want to be opt-in. The generated CLI doesn't do any threading so I don't think it should be the default.

Reply:
The users of the class never directly interact with the Dictionary instances.

Reply:
The API only requires `IEnumerable<KeyValuePair<...>>`
The dictionary instances are an implementation detail.

Reply:
why are those properties public then?

Reply:
The rationale for having them is in the code documentation.

Reply:
The point I'm trying to make here is that we don't need a concurrent structure since the host is going to be single threaded. And that we should simplify all of that to provide a cleaner API as well as reduce maintenance burden. The doc comments, while great, do not provide a scenario as off why anybody would need a concurrent data structure. Hence my original question/comment.

Reply:
The scenario was suggested [here](https://github.com/microsoft/kiota-cli-commons/pull/65#discussion_r1300722201)

Reply:
Thanks for the context. @peombwa I don't believe there's effectively a scenario where this might be used in a multi threaded context since this is for CLIs and a request will only ever be executed on a single thread. Can you expand on your thoughts with examples please?

Reply:
The multithreaded scenario I was thinking of is when CLI is used from a C# code via `Process.Start(...)`. One or more threads would run in the context of the process. By starting a process, you are implicitly creating at least one new thread, primary thread. Any thread can create additional threads.

Reply:
I don't know that we can guarantee no multi-threading. The graph CLI has that guarantee, but self-serve scenarios could do it differently if they wanted.

Reply:
But then again, a single request will only ever be executed by one thread at a time. And to guarantee not mixing the headers between requests in that scenario we'd either need some kind of request mapping code in the store (which we don't have) or to have a single instance of the store per request (which I believe is the way the middleware is supposed to operate, or at least we should strive for that as it's the simplest implementation)

Reply:
It's not the request I'm worried about. It's adding headers to the store that could be a problem. A user could add headers to the store in a PLINQ query or 2 tasks. In that situation, I think a concurrent dictionary would be useful to have without needing to create their own class.

Reply:
alright then, let's just use a ConcurrentDictionary, and have a single use mode for this store.

Reply:
I prefer giving users an option. A concurrent dictionary is very slow when compared to a normal dictionary. Otherwise, the graph CLI will have to add its own store.

Reply:
We're talking nanoseconds here....

Those singletons are not that useful in themselves, especially when you consider the consumer is likely to make multiple requests and don't want to share the headers between requests. We should probably remove those or convert them to a static factory (that doesn't provide much benefits over just use the constructor at this point)

Then why using a boolean flag instead of letting them pass the dictionary instance instead (IDictionary interface) so they have more control over the data type and over the behaviour (ignore case or not)?
Lastly you could add a default constructor that defaults to a regular dictionary with ordinal ignore case comparer.

Reply:
I think I should mention why performance for the CLI is important here. The CLI is a short-lived process that a user runs and expects an expedient response, especially if they don't expect it to be doing a lot of work. For Microsoft Graph, we are currently penalized on System.CommandLine startup time (uses up ~700ms), and on windows computers due to defender's scanning time (varies) which leaves us very little headroom for doing actual work in the CLI. That's why I'm careful on not doing unnecessary work at runtime.

Reply:
> especially when you consider the consumer is likely to make multiple requests and don't want to share the headers between requests

The API has a `Drain` function which clears the headers. The singletons are there for the most common scenario where users have 1 instance of the header store to make a single request. They need a way to get the correct instance of the store for the CLI middleware to use when adding headers.

The rationale behind using a boolean flag is there are only 2 scenarios we handle for. I can change the API to take in a dictionary interface which increases the leakiness of the abstraction a little, but I don't have a strong opinion on that. The only scenario I can think of that would be surprising is when a user passes in their own dictionary and doesn't give it a StringComparer. The behavior of the class will change complicating things even further.

Reply:
Thanks for the additional context. Coordinating the call to Drain in a multi-threaded context over network IO (the requests) to make sure the consumer has the right headers set when the request is being executed will simply be a nightmare. Let's remove this trap from the API surface (the singleton).

Since the comparer is not accessible from the IDictionary interface, I'd be ok with keeping a flag.

Reply:
What API would you suggest we provide instead for users to manage the lifetime of the headers store?

Reply:
I was going to do that initially, but I added the singleton because it's hard to address [this scenario](https://github.com/microsoft/kiota-cli-commons/pull/65#discussion_r1300720028) while asking people to manage the lifetime manually.

Reply:
why? the getter parameter could be named factory, be optional and have a default value of () => new InMemoryHeadersStore()

Reply:
Because the getter functor provided to `RegisterHeadersOption` should return the same instance as the one passed into `NativeHttpHeadersHandler` for it to work. Otherwise the headers collected from the CLI parser wouldn't be added to the HTTP request

Reply:
this is where the current design shows limitations. The lifecycle of the store has to be carefully managed by the caller. As opposed to relying on code gen (which has a size cost) where the lifecycle is managed by the generated code.

But again, this is related but not impossible to work with. You could simply:
1. instantiate a store, keep a reference to it
2. pass that reference to the command line middleware
3. pass that reference to the http middleware

Without the need for singletons as we've already established the lifecycle needs to be managed by the calling application multiple times.

Reply:
With the singleton, lifecycle is managed automatically. i.e. there's 1 instance. That works for most scenarios. It's more complicated scenarios that will need to manually manage the lifecycle. In that case, they can use an IoC container or a different kind of object registry.

Code generation, while simpler, carries a lot more than just a size cost. For example:
1. the size cost means windows defender scanning will be even slower
2. we'll lose every single extensibility option we've provided with this design like:
    - choosing the header format through overriding the parse logic.
    - changing the name of the `--header` option to work around generation conflicts
    - disabling header support if they don't need it

Also, all the complex scenarios we're discussing here would be very difficult to handle with generated code. I've mentioned some of these considerations in the issue tracking this feature.

Reply:
> That works for most scenarios. It's more complicated scenarios that will need to manually manage the lifecycle. In that case, they can use an IoC container or a different kind of object registry.

Hence the suggestion to remove the singletons to both cleanup the API surface and avoid leading users to make mistakes.

Reply:
That suggestion makes it more complicated for the simple cases.